File: docs/JAMIA/comparisons/RPython/TM/48719937_idxmax_object_dtype.py
# ID: SO#48719937
# Language: Python
# Bug class: Value type
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: idxmax() on object-dtype column fails. Numeric reduction on wrong type.
# Reproduction status: Fixed (pandas 2.x handles object-dtype idxmax)
# Type system catch: `number` column type required for numeric reductions

import pandas as pd
import numpy as np

# Simulate the bug: DataFrame with object-dtype numeric column
c_params = [0.01, 0.1, 1, 10, 100]
results = pd.DataFrame(index=range(len(c_params), 2), columns=['C_parameter', 'Mean recall score'])
results['C_parameter'] = c_params

# Assign recall scores as computed values
recall_scores = [0.95, 0.90, 0.92, 0.92, 0.92]
for j, score in enumerate(recall_scores):
    results.iloc[j, 1] = score

print(f"dtype of 'Mean recall score': {results['Mean recall score'].dtype}")
print(results)
print()

try:
    best = results.loc[results['Mean recall score'].idxmax()]['C_parameter']
    print(f"Best: {best}")
except TypeError as e:
    print(f"Crash: {e}")

File: docs/JAMIA/comparisons/RPython/TM/48719937_idxmax_object_dtype.ts
/**
 * RPython SO#48719937 — TypeError: reduction operation 'argmax' not allowed for this dtype
 * Effect: Crash
 * Bug class: Type coercion
 *
 * In pandas, idxmax() fails on object-dtype columns. The column contains
 * numeric values stored as strings due to DataFrame construction with wrong
 * index size (extra NaN rows coerce column to object).
 *
 * In tidy-ts, the column type is explicit. If the column is string,
 * numeric reductions are rejected at compile time.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const results = createDataFrame([
  { C_parameter: 0.01, mean_recall: "0.95" },
  { C_parameter: 0.1, mean_recall: "0.90" },
  { C_parameter: 1, mean_recall: "0.92" },
  { C_parameter: 10, mean_recall: "0.92" },
  { C_parameter: 100, mean_recall: "0.92" },
]);

// mean_recall is string — numeric reduction is rejected
// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.max(results.extract("mean_recall"));

// Fix: parse the column first
const parsed = results.mutate({ recall_num: (r) => parseFloat(r.mean_recall) });
console.log(s.max(parsed.extract("recall_num")));
